src/java/htsjdk/samtools/SamPairUtil.java

```diff
@@ -466,6 +467,7 @@ private void advance() {
         }
 
         public SAMRecord next() {
+            if(!hasNext()) { throw new NoSuchElementException(); }

```

Comment:
Thanks. Alec. As you know I'm not familiar with the history as much as you
are, which is why I asked. I agree that a closed iterator should not allow
any further reading, but it seems that it will even throw an exception on
hasNext(), which is what confused me.

On Sun, Feb 21, 2016 at 2:07 PM, alecw <notifications@github.com> wrote:

> In src/java/htsjdk/samtools/SamPairUtil.java
> <https://github.com/samtools/htsjdk/pull/480#discussion_r53571266>:
>
> > @@ -466,6 +467,7 @@ private void advance() {
> >          }
> >
> >          public SAMRecord next() {
> > +            if(!hasNext()) { throw new NoSuchElementException(); }
>
> Hi Yossi, OK, I assume you're referring to
> AbstractBamIterator.assertOpen(). Let me preface the following by saying
> that reasonable people can disagree on this question. The behavior of
> calling next() on an exhausted iterator is clear from the Iterator
> interface -- it should throw NoSuchElementException. The question is: is a
> closed iterator an exhausted iterator? I would argue that it is not. I
> would further argue that it is a bug to call next() or hastNext() on a
> closed iterator. An analogy would be calling a read() method on a Reader or
> InputStream that has been closed. I wouldn't expect that I would get a
> result that indicated EOF; rather I would expect to get an exception
> indicating that I tried to read from a closed InputStream. When we started
> writing Picard, part of our design philosophy was to be strict and throw an
> exception at anything that smelled funny. Not everyone agrees with this
> philosophy, and over the years we have received numerous complaints that
> samtools accepts something that Picard does not, and many times our
> response was that we were comfortable with the strict behavior. Just my
> $0.02.
> … <#1078751382_>
> -Alec
> On Sat, Feb 20, 2016 at 6:06 PM, Yossi Farjoun <notifications@github.com>
> wrote: In src/java/htsjdk/samtools/SamPairUtil.java <#480 (comment)
> <https://github.com/samtools/htsjdk/pull/480#discussion_r53559011>>: > @@
> -466,6 +467,7 @@ private void advance() { > } > > public SAMRecord next() {
> > + if(!hasNext()) { throw new NoSuchElementException(); } Sorry @alecw
> <https://github.com/alecw> <https://github.com/alecw> I wasn't clear
> enough. I meant the assertOpen() in htsjdk.samtools.SAMRecord#hasNext() and
> htsjdk.samtools.SAMRecord#next() shouldn't hasNext() simply return false if
> the iterator is closed, not throw an exception? And the exception being
> thrown in this case should be a NoSuchElementException rather than a
> AssertionError, no? The only reason I mentioned you here was that you
> seemed to have added the AbstractBamIterator about two years ago, and
> together with that change, added the calls to assertOpen. — Reply to this
> email directly or view it on GitHub <
> https://github.com/samtools/htsjdk/pull/480/files#r53559011>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/480/files#r53571266>.
>